### DISCUSSION TITLE
Observe C(R)UD events

### DISCUSSION THREAD

**Discussion by MartinP7r** [groue/GRDB.swift#1285] [Q&A] [answered]
Hello 👋🏼 

With [`ValueObservation`](https://github.com/groue/GRDB.swift#valueobservation) we can track changes to the database based on a tracked value.  

> Tracked changes are insertions, updates, and deletions
  
Is there a way to observe insert/delete/update of rows directly and get back the objects that were inserted/deleted, e.g. as a `PassthroughSubject`?  

I'm imagining an API along the lines of

```swift
ValueObservation(events: [.deletion])
  .tracking(Player.self)
  .publisher(in: db.reader)
```

If it's already available and I overlooked it in the docs I'm very sorry!

**Comment by groue** [comment]
Hello @MartinP7r,

You can turn a ValueObservation into a Combine publisher, with the same api as the one you are imagining.

For details, see [GRDB ❤️ Combine](https://github.com/groue/GRDB.swift/blob/master/Documentation/Combine.md#database-observation).

For more usages of ValueObservation (including Combine, but also asynchronous sequences), see [ValueObservation Usage](https://github.com/groue/GRDB.swift/blob/master/README.md#valueobservation-usage).

**Comment by groue** [comment, marked as answer]
Oops, my initial answer is off topic.

> Is there a way to observe insert/delete/update of rows directly and get back the objects that were inserted/deleted.

You can build this from the low-level [TransactionObserver protocol](https://github.com/groue/GRDB.swift/blob/master/README.md#transactionobserver-protocol). You'll get there the most information that SQLite can provide about database changes.

**Reply by MartinP7r** [reply]
Oh, thank you! Sorry if my question was a little hard to understand.  
I'll have a look at `TransactionObserver`!

**Reply by groue** [reply]
😬 SQLite won't let you fetch the deleted items. It is a programmer error to access the database as SQLite is notifying database changes with the [notification callback](https://www.sqlite.org/c3ref/update_hook.html) that `TransactionObserver` relies on. This means that `TransactionObserver` is notified of deletions, but it must not fetch the deleted row(s).

Maybe [Support for SQLite Pre-Update Hooks](https://github.com/groue/GRDB.swift/blob/master/README.md#support-for-sqlite-pre-update-hooks) can help, but it's not always available.

**Reply by groue** [reply]
You have other options to evaluate:

- [`deleteAndFetchAll`](https://github.com/groue/GRDB.swift/blob/master/README.md#persistence-methods-and-the-returning-clause) (new in GRDB 6) which deletes, and returns the deleted rows.
- A [trigger](https://www.sqlite.org/lang_createtrigger.html) that populates a table of deleted rows, which you can observe.

**Reply by MartinP7r** [reply]
You're right, this doesn't help me as I'm trying to observe deleted items from a separate part of the app before discarding them.  
Conceptually trigger or a pre-update hook seem to be my best options here.  
I any case, thank you for showing me all these different paths of possible solutions! 🤗

**Reply by groue** [reply]
All right :) I'm curious about the solution that will work for you!

**Reply by funct7** [reply]
I'm assuming that the OP wanted to receive only the "delta" not the end result. I think it might be a good extension to provide for people migrating from Realm.

**Reply by groue** [reply]
All contributions are welcome, @funct7. You might be interested in https://github.com/groue/GRDB.swift/discussions/1523 as well.
